### ISSUE TITLE
how to make self-sampled datasets

### ISSUE THREAD

**Issue by SnallQiu** [f3rm/f3rm#11]
Hi, I am new to nerfstudio. If i already have a calibrated camera and the corresponding robot arm, and get 50 pictures in a scenario, how to generate a transform.json and nerf_to_world.json?

At first I thought it was to use `ns-process-data images --data datasets/f3rm/panda/scene_001/images/ --output-dir XXX` directly, but the result transform.json is not the same as yours in `datasets/f3rm/panda/scene_001/transforms.json`, I think this process seems to calculate the calibration within the pictures not the one we need to provide that we calibrated before.

**Comment by williamshen-nz** [comment]
If you have a calibrated eye-in-hand camera you can use forward kinematics on the robot to get the camera poses. Then you can create your own `transform.json` format that follows the format Nerfstudio expects. Note that Nerfstudio uses the OpenGL coordinate convention (https://docs.nerf.studio/quickstart/data_conventions.html#coordinate-conventions).

Using `ns-process-data images ...` will run COLMAP on the images to get the camera poses which are in an arbitrary coordinate frame, but since you already have calibrated the camera you can just use that.

When you train a NeRF and feature field, by default a scaling, rotation and translation is applied to move the cameras within a bounding box (in `dataparser_transforms.json`). You can compute the inverse transformation and add it to `nerf_to_world.json` (see https://github.com/f3rm/f3rm/issues/7).

**Comment by SnallQiu** [comment]
Hi, thanks for your quick reply!
I still have some questions while I'm using ns-train in my datasets.
Now, I have collected 50 pngs with the corresponding camera pose(to the world frame). I tried to make the transforms.json and train a  DFF, but the output feature field is messy.

I know the camera coordinate and the OpenGL coordinate are exactly not similar(  Camera coordinate, inverse y,z axis -> OpenGL coordinate), and I have made coordinate convention while forming the transforms.json, is my code right?

```
    def _get_camera_trans(self):
        trans = self.tf_buffer.lookup_transform("base_link", # world frame
                                                "camera_link",  # camera frame
                                                rospy.Time(0),
                                                rospy.Duration(1.0))

        rot = trans.transform.rotation
        # camera
        rot_matrix = tf.transformations.quaternion_matrix([rot.x, rot.y, rot.z, rot.w])[:3, :3]
        # nerf
        rot_matrix_nerf = rot_matrix
        col2 = rot_matrix_nerf[:, 1].copy()
        col3 = rot_matrix_nerf[:, 2].copy()
        # Camera coordinate, inverse y,z axis -> OpenGL coordinate 
        rot_matrix_nerf[:, 1] = -col2
        rot_matrix_nerf[:, 2] = -col3

        transl = trans.transform.translation
        x = transl.x
        y = transl.y
        z = transl.z

        matrix = np.eye(4)
        matrix[:3, :3] = rot_matrix_nerf
        matrix[:3, 3] = np.array([x, y, z])


        return matrix

matrix = self._get_camera_trans()
self.frames.append(
                {
                    "file_path": 'images/image_{}.png'.format(self.image_cnt),
                    "transform_matrix": matrix.tolist()
                }
)
```

However, using `ns-process-data` to deal with this 50png first(get a transforms.json), and use ns-train to train a DFF, the result is good. I'm puzzled about the world frame is not fixed when trying to record object demonstration.

In summary, a set of pictures taken by a camera with calibrated external params, how to correctly form a transforms.json according to the requirements of nerfstudio?

Hopes for your reply again, thank you!

**Comment by williamshen-nz** [comment]
Here is the code I use to convert between OpenCV and OpenGL conventions:

```
import numpy as np


def convert_opengl_to_opencv(c2w: np.ndarray):
    """Convert from OpenGL to OpenCV coordinate convention."""
    flip_yz = np.eye(4)
    flip_yz[1, 1] = -1
    flip_yz[2, 2] = -1
    c2w = np.matmul(c2w, flip_yz)
    return c2w


def convert_opencv_to_opengl(c2w: np.ndarray):
    """Convert from OpenCV to OpenGL coordinate convention."""
    flip_yz = np.eye(4)
    flip_yz[1, 1] = -1
    flip_yz[2, 2] = -1
    flip_yz_inv = np.linalg.inv(flip_yz)
    c2w = np.matmul(c2w, flip_yz_inv)
    return c2w
```

Assuming your calibration is accurate and you represent the poses in the right convention, you should be able to get a good result. Don't forget about setting intrinsics and any distortion parameters as well that may be relevant.

I suggest that you check the poses look correct in the Nerfstudio viewer during training, or visually compare your calibrated poses with the COLMAP poses from `ns-process-data` and see if they look right.

I'm not sure what you mean by the feature field looking 'messy'. If things do not look sharp, that might be due to inaccurate calibration.

**Comment by SnallQiu** [comment]
Thanks for your detailed explanation!
Finally, I found that my camera’s internal calibrated params are not accurate. 
My robot is kinova gen3 with a vision module. The kinova-vision package has provided the internal calibrated params, but the camera has a "Auto focus function" which will change the internal calibrated params in real time.  And it can be closed in the kinova localhost web, if someone else looks this issue, maybe helpful.

The next question is that we need to record the object demonstration.
In your datasets of `datasets/f3rm/panda_demos/caterpillar/scene_demo.json`, is the grasping pose here in the world coordinate?

Hopes for your reply again, thank you!

**Comment by williamshen-nz** [comment]
Yes, the grasping pose is specified in the world frame. See:

- https://github.com/f3rm/f3rm/blob/main/assets/datasets.md#panda_demos
- https://github.com/f3rm/f3rm/tree/main/f3rm_robot#how-can-i-generate-my-own-task-embeddings

Regarding your camera, you can try disabling autofocus.

**Comment by SnallQiu** [comment]
I'm sorry for missing the above readme. Thank you for your reply!

In the later tests, we find that disabling autofocus and making transforms.json with self-calibrated camera params, the result is still messy, see:
![image](https://github.com/f3rm/f3rm/assets/33489953/a78fc5ce-c970-4203-8d71-3ca7898b8114)

The COLMAP result is :
![image](https://github.com/f3rm/f3rm/assets/33489953/d43c6104-3bd1-4507-88cd-cb5190be6b71)


The camera pose in the localhost viewer seems right but a little shifted in some areas compared to the COLMAP result, I'm still finding the detailed error. And I would like to use self-calibrated camera poses rather than the COLMAP calibrated camera poses, the reason is that we can fix the world frame in the robot base_link which is convenient when recording the object scanning and the grasping pose.

In your paper appendix, I see you use the COLMAP to calibrate the camera pose in the scene scanning while placing some markers to convert the COLMAP frame to the world frame. In my test, the COLMAP frame will be changed even by fixing the everytime scanning trajectory. Unless we put the markers in each scanning step, we can convert the world frames between different COLMAP result.

And I'm wondering do you maintain the same world frame in the scene scanning and **the object scanning**?  It seems no markers in your object scanning. 
If I choose to use the COLMAP camera pose, my question is that do I need to keep the world coordinate of scene scanning consistent with the world coordinate of object scanning or make a convention between the two world frame?

Hopes for your reply again, thank you!

**Comment by williamshen-nz** [comment]
If your camera supports RGB-D, I'd suggest taking the multi-view RGB-D images and projecting all the point clouds with your self-calibrated poses. If they align then there is something likely going wrong going on the NeRF at which point you will need to further debug.

> And I'm wondering do you maintain the same world frame in the scene scanning and the object scanning?

We run the same robot joint position trajectory every time we scan, hence we can reuse the trajectories we get from COLMAP for which we compute the transform to the world frame.

I don't know what you mean by scene and object scanning, we just have a dedicated scan for calibration in which we put lots of objects and markers on the table so COLMAP can find good SIFT features.

**Comment by SnallQiu** [comment]
Thank you, I got your idea. Use COLMAP for calibration once, and use a known pose marker for the conversion of world frame. Then use the fixed calibration camera pose each scan time to form the transform.json .

And can you show the code about how to transform the dataparser_transforms.json to the nerf_to_world.json? Maybe my code has some wrong when transforming the scale. My result in  nerf_to_world.json is a bit large and can not get any optimize step when try to optimize.
terminal output like this:
`Sampling point cloud for visualization:   0%|                                                                                                       0/200000 [1:24:22<?, ?it/s
`

Besides, is my scale value(**about 0.2**) in the dataparser_transforms.json seems normal?


Hopes for your reply again, thank you!

**Comment by SnallQiu** [comment]
> My result in nerf_to_world.json is a bit large and can not get any optimize step when try to optimize.

I read your source code and find the reason is that our world frame not exactly in the table, then the bounding box remove almost every points.

**Comment by williamshen-nz** [comment]
I recommend you visualize the point cloud sampled in the NeRF coordinate frame (you can use Nerfstudio for this: https://docs.nerf.studio/quickstart/export_geometry.html). Then you can apply your `nerf_to_world` transform and see if it matches what you expect.

We cannot tell you if your scale is correct. It differs based on the poses and how you scanned.

The sampling point cloud hangs as there are probably no points in the workspace bounding box. You need to change these parameters to match your workspace. In our case, the world frame is anchored at the base link of the Panda robot.

https://github.com/f3rm/f3rm/blob/main/f3rm_robot/args.py#L34-L36
